pacman/model/graphs/machine/machine_vertex.py

```diff
@@ -16,28 +16,88 @@
 from six import add_metaclass
 from spinn_utilities.abstract_base import AbstractBase, abstractproperty
 from pacman.model.graphs import AbstractVertex
+from pacman.model.graphs.common import Slice
+from pacman.model.graphs.application import ApplicationVertex
+from pacman.exceptions import PacmanInvalidParameterException
 
 
 @add_metaclass(AbstractBase)
 class MachineVertex(AbstractVertex):
     """ A machine graph vertex.
     """
 
-    __slots__ = []
+    __slots__ = ["_app_vertex", "_index", "_vertex_slice"]
+    _DEFAULT_SLICE = Slice(0, 0)
 
-    def __init__(self, label=None, constraints=None):
+    def __init__(self, label=None, constraints=None, app_vertex=None,
+                 vertex_slice=None):
         """
         :param label: The optional name of the vertex
         :type label: str or None
         :param iterable(AbstractConstraint) constraints:
             The optional initial constraints of the vertex
+        :param app_vertex:
+            The application vertex that caused this machine vertex to be
+            created. If None, there is no such application vertex.
+        :type app_vertex: ApplicationVertex or None
+        :param vertex_slice:
+            The slice of the application vertex that this machine vertex
+            implements.
+        :type vertex_slice: Slice or None
         :raise PacmanInvalidParameterException:
             If one of the constraints is not valid
         """
         if label is None:
             label = str(type(self))
         super(MachineVertex, self).__init__(label, constraints)
         self._added_to_graph = False
+        self._app_vertex = app_vertex
+        self._index = None
+        if app_vertex is not None and not isinstance(
+                app_vertex, ApplicationVertex):
+            raise PacmanInvalidParameterException(
+                "app_vertex", app_vertex, "must be an application vertex")
+        if vertex_slice is not None:
```

Comment:
I think remember_associated_machine_vertex should be called from here rather than in different places.

Reply:
If that works with doing graph filtering, yes.

Reply:
I've instead made it be done when the vertex is added to the graph. There was only one case where it wasn't previously the step immediately after adding to the graph, and that was a case (LPG machine vertex creation) where changing to be that way was fine.